crates/sp-lightclient/src/lib.rs

```diff
@@ -391,4 +404,99 @@ impl<Header: HeaderT, Store: Storage<Header>> HeaderImporter<Header, Store> {
 
         Ok(())
     }
+
+    /// Finalize the header at K-depth from the best block and prune remaining forks at that number.
+    /// We want to finalize the header from the current finalized header until the K-depth number of the best.
+    /// 1. In an ideal scenario, the current finalized head is one number less than number to be finalized.
+    /// 2. If there was a re-org to longer chain when new header was imported, we do not want to miss
+    ///    pruning fork headers between current and to be finalized number. So we go number by number and prune fork headers.
+    /// 3. If there was a re-org to a shorter chain and to be finalized header was below the current finalized head,
+    ///    fail and let user know.
+    fn finalize_header_at_k_depth(store: &mut Store) -> Result<(), ImportError<Header>> {
+        let k_depth = store.chain_constants().k_depth;
+        let current_finalized_header = store.finalized_header();
+
+        // ensure we have imported at least K-depth number of headers
+        let number_to_finalize = match store.best_header().header.number().checked_sub(&k_depth) {
+            // we have not progressed that far to finalize yet
+            None => {
+                // if the chain re-org happened to smaller chain and if there was any finalized heads,
+                // fail and let the user decide what to do
+                if *current_finalized_header.header.number() > Zero::zero() {
+                    return Err(ImportError::SwitchedToForkBelowArchivingDepth);
+                }
+
+                return Ok(());
+            }
+
+            Some(number) => number,
+        };
+
+        match number_to_finalize.cmp(current_finalized_header.header.number()) {
+            Ordering::Less => Err(ImportError::SwitchedToForkBelowArchivingDepth),
+            // nothing to do as we finalized the header already
+            Ordering::Equal => Ok(()),
+            // finalize heads one after the other and prune any forks
+            Ordering::Greater => {
+                let mut current_finalized_number = *current_finalized_header.header.number();
+
+                while current_finalized_number < number_to_finalize {
+                    let number_to_finalize = current_finalized_number
```

Comment:
A bit confusing that you're reusing outer `number_to_finalize` name internally, I think you could have just incremented `current_finalized_number` here instead of at the end of the loop with the same result:
```suggestion
                    current_finalized_number = current_finalized_number
```